File: docs/test-cases/ssrf.go
// SSRF — intentionally vulnerable. DO NOT deploy.
package main

import (
	"fmt"
	"io"
	"net/http"
)

// BUG: fetches any URL the caller supplies
func previewHandler(w http.ResponseWriter, r *http.Request) {
	url := r.URL.Query().Get("url")
	// No validation — attacker can reach http://169.254.169.254/latest/meta-data/
	resp, err := http.Get(url)
	if err != nil {
		http.Error(w, err.Error(), 500)
		return
	}
	defer resp.Body.Close()
	body, _ := io.ReadAll(resp.Body)
	fmt.Fprintf(w, "Status: %d\nBody: %s", resp.StatusCode, body[:500])
}

// BUG: proxy endpoint with no host validation
func proxyHandler(w http.ResponseWriter, r *http.Request) {
	target := r.URL.Query().Get("target")
	// Attacker sends target=http://localhost:6379/ to reach Redis
	resp, _ := http.Get(target)
	defer resp.Body.Close()
	io.Copy(w, resp.Body)
}

func main() {
	http.HandleFunc("/preview", previewHandler)
	http.HandleFunc("/proxy", proxyHandler)
	http.ListenAndServe(":8080", nil)
}

File: docs/test-cases/ssrf.rs
// SSRF — intentionally vulnerable. DO NOT deploy.
use actix_web::{get, web, App, HttpResponse, HttpServer};

// BUG: fetches any URL the caller supplies
#[get("/preview")]
async fn preview(query: web::Query<PreviewQuery>) -> HttpResponse {
    // No validation — attacker can reach http://169.254.169.254/
    let resp = reqwest::get(&query.url).await.unwrap();
    let body = resp.text().await.unwrap();
    HttpResponse::Ok().body(format!("Fetched: {}", &body[..500.min(body.len())]))
}

// BUG: proxy with no host validation or IP blocking
#[get("/proxy")]
async fn proxy(query: web::Query<ProxyQuery>) -> HttpResponse {
    // Attacker sends target=http://localhost:6379/ to probe internal services
    let client = reqwest::Client::builder()
        .redirect(reqwest::redirect::Policy::limited(10)) // follows redirects
        .build()
        .unwrap();
    let resp = client.get(&query.target).send().await.unwrap();
    let body = resp.bytes().await.unwrap();
    HttpResponse::Ok().body(body)
}

#[derive(serde::Deserialize)]
struct PreviewQuery { url: String }

#[derive(serde::Deserialize)]
struct ProxyQuery { target: String }

#[actix_web::main]
async fn main() -> std::io::Result<()> {
    HttpServer::new(|| App::new().service(preview).service(proxy))
        .bind("0.0.0.0:8080")?
        .run()
        .await
}
